specs/experimental/op-stack-manager.md

```diff
@@ -0,0 +1,261 @@
+
+```solidity
+struct SystemConfigInputs {
+  address systemConfigOwner;
+  uint256 overhead;
+  uint256 scalar;
+  bytes32 batcherHash;
+  address unsafeBlockSigner;
+}
+```
+
+Details on each parameter can be found in the code comments of the [`SystemConfig` contract].
+Note that not all `SystemConfig` parameters are able to be configured.
+This is because the OP Stack Manager enforces compliance with the [standard configuration], and only allows
+configuration where allowed by the standard configuration.
+
+The `L2OutputOracleInputs` struct is defined as follows:
+
+```solidity
+struct L2OutputOracleInputs {
+  address proposer;
+  address challenger;
+}
+```
+
+Details on each parameter can be found in the code comments of the [`L2OutputOracle` contract].
+Note that not all `L2OutputOracle` parameters are able to be configured, for the reason as the `SystemConfig`.
+
+On success, the following event is emitted:
+
+```solidity
+event Deployed(uint256 indexed l2ChainId, SystemConfig indexed systemConfig);
+```
+
+This method reverts on failure. This occurs when:
+
+- The input `l2ChainId` does not comply with the restrictions above.
+- The resulting configuration is not compliant with the [standard configuration].
+
+### Getter Methods
+
+The following interface defines the available getter methods:
+
+```solidity
+/// @notice The logic address and initializer selector for an implementation contract.
+struct Implementation {
+  address logic; // Address containing the deployed logic contract.
+  bytes4 initializer; // Function selector for the initializer.
+}
+
+/// @notice Returns the latest approved release of the OP Stack contracts.
+/// @notice Release strings follow semver and are named with the
+/// format `op-contracts/vX.Y.Z`.
+function latestRelease() external view returns (bytes32);
+
+/// @notice Maps a release version to a contract name to its implementation data.
+function implementation(
+  bytes32 release,
+  bytes32 contractName
+) external view returns (Implementation memory);
+
+/// @notice Maps an L2 Chain ID to the SystemConfig address for that chain.
+/// @notice Most information for a chain can be found from it's the SystemConfig,
+/// with ProtocolVersions being the exception. ProtocolVersions is a singleton
+/// for all chains deployed by this contract.
+function systemConfig(uint256 chainId) external view returns (SystemConfig);
+```
+
+## Implementation
+
+### Batch Inbox Address
+
+Limiting the L2 chain ID to a `uint64` allows the chain's [Batch Inbox] address
+to be computed and set by the OP Stack Manager at `deploy` time. The batch
+inbox address follows the convention `0xFF000...000{l2ChainId}`. Specifically,
+it is computed as follows:
+
+```solidity
+/// @notice Maps an L2 chain ID to an L1 batch inbox address of the
+/// form `0xFF000...000{chainId}`.
+function chainIdToBatchInboxAddress(uint256 l2ChainId) public pure returns (address) {
+    return address(uint160(0xFF) << 152 | uint160(l2ChainId));
+}
+```
```

Comment:
I like the idea of having a default batch inbox address that is autogenerated by the chain id, but I think eventually it will need to be dynamic if we ever want something like https://github.com/ethereum-optimism/specs/issues/221 to be able to work

Also there may exist a standard chain today that doesn't follow this schema. So what i mean is that i think its not important to preserve the property that given an L2 chain id, you can compute the batch inbox. The system config + superchain config should be the anchor points for as much information about L2 chains as possible

Reply:
> The system config + superchain config should be the anchor points for as much information about L2 chains as possible

Ultimately I agree with this and am ok with not preserving the existing property. For this version of OP Stack Manager though, we either need to:
1. Have some way (such as the current one) of deriving batch inbox address from the given inputs, or
2. Add batch inbox address as an input.

I went with (1) to simplify the deploy process by removing one input, which also minimizes the chances of user error.

Are you ok with sticking with this approach for the first version, or do you have a preferred alternative?

Reply:
I am fine with sticking with this approach for the first version, its fine to say `uint64` only for now, altho i think we should solve #94 in a dumb way eventually and not overthink it

Reply:
You're right we shouldn't overthink #94—I've put a proposal in https://github.com/ethereum-optimism/specs/issues/94#issuecomment-2176212325. Let's see if we can get that change finalized this week. If we can't, then we can stick with this approach for the first version

Reply:
Optimistically changes to use the proposed 256-bit approach in c2d0835

Reply:
Resolving since #94 has been merged